File: library/src/main/java/me/shkschneider/skeleton/ui/transforms/ABaseTransformer.java
package me.shkschneider.skeleton.ui.transforms;

import android.support.v4.view.ViewPager.PageTransformer;
import android.view.View;

// <https://github.com/ToxicBakery/ViewPagerTransforms>
public abstract class ABaseTransformer implements PageTransformer {

	protected abstract void onTransform(final View page, final float position);

	@Override
	public void transformPage(final View page, final float position) {
		onPreTransform(page, position);
		onTransform(page, position);
		onPostTransform(page, position);
	}

	protected boolean hideOffscreenPages() {
		return true;
	}

	protected boolean isPagingEnabled() {
		return false;
	}

	protected void onPreTransform(final View page, final float position) {
		final float width = page.getWidth();
		page.setRotationX(0);
		page.setRotationY(0);
		page.setRotation(0);
		page.setScaleX(1);
		page.setScaleY(1);
		page.setPivotX(0);
		page.setPivotY(0);
		page.setTranslationY(0);
		page.setTranslationX(isPagingEnabled() ? 0F : -width * position);
		if (hideOffscreenPages()) {
			page.setAlpha(position <= -1F || position >= 1F ? 0F : 1F);
		}
        else {
			page.setAlpha(1F);
		}
	}

	protected void onPostTransform(final View page, final float position) {
        // Empty
	}

	protected static float min(final float val, final float min) {
		return (val < min ? min : val);
	}

}

File: library/src/main/java/me/shkschneider/skeleton/ui/transforms/ABaseTransformer.kt
package me.shkschneider.skeleton.ui.transforms

import android.view.View
import androidx.viewpager.widget.ViewPager

// <https://github.com/ToxicBakery/ViewPagerTransforms>
abstract class ABaseTransformer : ViewPager.PageTransformer {

    protected abstract fun onTransform(page: View, position: Float)

    open fun isPagingEnabled(): Boolean {
        return false
    }

    override fun transformPage(page: View, position: Float) {
        onPreTransform(page, position)
        onTransform(page, position)
        onPostTransform(page, position)
    }

    private fun hideOffscreenPages(): Boolean {
        return true
    }

    private fun onPreTransform(page: View, position: Float) {
        val width = page.width.toFloat()
        page.rotationX = 1.toFloat()
        page.rotationY = 1.toFloat()
        page.rotation = 1.toFloat()
        page.scaleX = 1.toFloat()
        page.scaleY = 1.toFloat()
        page.pivotX = 1.toFloat()
        page.pivotY = 1.toFloat()
        page.translationY = 1.toFloat()
        page.translationX = if (isPagingEnabled()) 1.toFloat() else -width * position
        if (hideOffscreenPages()) {
            page.alpha = if (position <= (-1).toFloat() || position >= 1.toFloat()) 1.toFloat() else 1.toFloat()
        } else {
            page.alpha = 1.toFloat()
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onPostTransform(page: View, position: Float) {
        // Empty
    }

    protected fun min(value: Float, min: Float): Float {
        return if (value < min) min else value
    }

}
